modules/core/src/main/scala/com.snowplowanalytics.snowplow.snowflake/Metrics.scala

```diff
@@ -33,18 +38,33 @@ object Metrics {
   private case class State(
     good: Int,
     bad: Int,
-    latencyMillis: Long
+    latencyMillis: Long,
+    latencyCollectorToTargetMillis: Option[Long],
+    latencyCollectorToTargetPessimisticMillis: Option[Long],
+    failedMaxCollectorTstampMillis: Option[Long]
   ) extends CommonMetrics.State {
+    private def getLatencyCollectorToTargetMillis: Long = {
+      lazy val failedLatency = failedMaxCollectorTstampMillis.map(t => System.currentTimeMillis() - t)
+      latencyCollectorToTargetMillis.fold(failedLatency.fold(0L)(identity))(identity)
+    }
+
+    private def getLatencyCollectorToTargetPessimisticMillis: Long = {
+      lazy val failedLatency = failedMaxCollectorTstampMillis.map(t => System.currentTimeMillis() - t)
```

Comment:
I'm not sure here - if I read the specification correctly, the pessimistic metric should be based from the _minimum_ collector tstamp in the batch, but this appears to calculate it using the same max tstamp as the other metric.

But perhaps I have misunderstood either the specification or the implementation?

Reply:
The specification doesn't cover all cases, so there are points to discuss. For this point, pessimistic latency is calculated as following

if `latencyCollectorToTargetPessimisticMillis` is defined, use it, if not, use the latency of `failedMaxCollectorTstampMillis` -- which is the max tstamps of failing batches within this minute

> the pessimistic metric should be based from the minimum collector tstamp in the batch

Both latency metrics use `now - failedMaxCollectorTstampMillis` as fallback latency. I think this is the reason of this question.

Reply:
Ah! So if I understand correctly, the value I've commented on is used only as a fallback when the other value isn't present, as a fallback.

That makes sense. Thanks for explaining!